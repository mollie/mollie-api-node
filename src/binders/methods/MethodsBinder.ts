import type TransformingNetworkClient from '../../communication/TransformingNetworkClient';
import type Method from '../../data/methods/Method';
import { type MethodData } from '../../data/methods/data';
import alias from '../../plumbing/alias';
import renege from '../../plumbing/renege';
import type Callback from '../../types/Callback';
import Binder from '../Binder';
import { type GetParameters, type ListParameters } from './parameters';

const pathSegment = 'methods';

export default class MethodsBinder extends Binder<MethodData, Method> {
  constructor(protected readonly networkClient: TransformingNetworkClient) {
    super();
    alias(this, { list: ['all', 'page'] });
  }

  /**
   * Retrieve a single method by its ID. Note that if a method is not available on the website profile a status `404 Not found` is returned. When the method is not enabled, a status `403 Forbidden` is
   * returned. You can enable payments methods via the Enable payment method endpoint in the Profiles API, or via your [Mollie Dashboard](https://www.mollie.com/dashboard).
   *
   * If you do not know the method's ID, you can use the methods list endpoint to retrieve all payment methods that are available.
   *
   * Additionally, it is possible to check if Wallets such as Apple Pay are enabled by passing the wallet ID (`applepay`) as the method ID.
   *
   * @since 2.0.0
   * @see https://docs.mollie.com/reference/v2/methods-api/get-method
   */
  public get(id: string, parameters?: GetParameters): Promise<Method>;
  public get(id: string, parameters: GetParameters, callback: Callback<Method>): void;
  public get(id: string, parameters?: GetParameters) {
    if (renege(this, this.get, ...arguments)) return;
    return this.networkClient.get(`${pathSegment}/${id}`, parameters);
  }

  /**
   * Retrieve all enabled payment methods. The results are not paginated.
   *
   * -   For test mode, payment methods are returned that are enabled in the Dashboard (or the activation is pending).
   * -   For live mode, payment methods are returned that have been activated on your account and have been enabled in the Dashboard.
   *
   * New payment methods can be activated via the Enable payment method endpoint in the Profiles API.
   *
   * When using the `first` sequence type, methods will be returned if they can be used as a first payment in a recurring sequence and if they are enabled in the Dashboard.
   *
   * When using the `recurring` sequence type, payment methods that can be used for recurring payments or subscriptions will be returned. Enabling / disabling methods in the dashboard does not affect
   * how they can be used for recurring payments.
   *
   * @since 3.0.0
   * @see https://docs.mollie.com/reference/v2/methods-api/list-methods
   */
  public list(parameters?: ListParameters): Promise<Method[]>;
  public list(parameters: ListParameters, callback: Callback<Method[]>): void;
  public list(parameters: ListParameters = {}) {
    if (renege(this, this.list, ...arguments)) return;
    return this.networkClient.list<MethodData, Method>(pathSegment, 'methods', parameters);
  }
}
