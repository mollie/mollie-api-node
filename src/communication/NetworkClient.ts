import https from 'https';
import fetch, { type RequestInit } from 'node-fetch';
import { type SecureContextOptions } from 'tls';

import { run } from 'ruply';
import type Page from '../data/page/Page';
import ApiError from '../errors/ApiError';
import type Options from '../Options';
import fling from '../plumbing/fling';
import DemandingIterator from '../plumbing/iteration/DemandingIterator';
import HelpfulIterator from '../plumbing/iteration/HelpfulIterator';
import Throttler from '../plumbing/Throttler';
import type Maybe from '../types/Maybe';
import type MaybeArray from '../types/MaybeArray';
import { type IdempotencyParameter } from '../types/parameters';
import breakUrl from './breakUrl';
import buildUrl, { type SearchParameters } from './buildUrl';
import dromedaryCase from './dromedaryCase';
import { idempotencyHeaderName, ResponseWithIdempotencyKey, retryingFetch } from './makeRetrying';
// The following line is only necessary for Node.js < 10.0.0, which we only secretly support. Should we ever drop that support completely, we can remove this import.
import { URL } from 'url';

/**
 * Like `[].map` but with support for non-array inputs, in which case this function behaves as if an array was passed
 * with the input as its sole element.
 */
function map<T, U>(input: Maybe<MaybeArray<T>>, callback: (value: T, index: number) => U, context?: any): U[] {
  if (Array.isArray(input)) {
    return input.map(callback, context);
  }
  if (input != undefined) {
    return [callback.call(context, input, 0)];
  }
  return [];
}

/**
 * Composes a `User-Agent` header value which looks something like
 * `'Node/10.0.0 Mollie/3.0.0 RockenbergCommerce/1.16.0'`.
 */
function composeUserAgent(nodeVersion: string, libraryVersion: string, versionStrings: Options['versionStrings']) {
  return [
    `Node/${nodeVersion}`,
    `Mollie/${libraryVersion}`,
    ...map(versionStrings, versionString => {
      const matches =
        // platform/version
        /^([^\/]+)\/([^\/\s]+)$/.exec(versionString) ??
        fling(() => {
          if (-1 == versionString.indexOf('/') || versionString.indexOf('/') != versionString.lastIndexOf('/')) {
            return new Error('Invalid version string. It needs to consist of a name and version separated by a forward slash, e.g. RockenbergCommerce/3.1.12');
          }
          return new Error('Invalid version string. The version may not contain any whitespace.');
        });
      const platform = dromedaryCase(matches[1]);
      const version = matches[2];
      return `${platform}/${version}`;
    }),
  ].join(' ');
}

/**
 * Throws an API error based on the passed cause.
 */
const throwApiError = run(
  /**
   * Returns whether the passed value is an object with a property with the passed name (`true`) or not (`false`).
   */
  function findProperty<K extends string>(value: unknown, name: K): value is Record<K, unknown> {
    return typeof value == 'object' && value != null && name in value;
  },
  findProperty =>
    function throwApiError(cause: unknown) {
      throw new ApiError(findProperty(cause, 'message') ? String(cause.message) : 'An unknown error has occurred');
    },
);

/**
 * Checks whether an API error needs to be thrown based on the passed result.
 */
async function processFetchResponse(response: ResponseWithIdempotencyKey) {
  // Request was successful, but no content was returned.
  if (response.status == 204) return true;
  // Request was successful and content was returned.
  const body = await response.json();
  if (Math.floor(response.status / 100) == 2) {
    return body;
  }
  // Request was not successful, but the response body contains an error message.
  if (null != body) {
    throw ApiError.createFromResponse(body, response.idempotencyKey);
  }
  // Request was not successful.
  throw new ApiError('An unknown error has occurred');
}

interface Data {}
interface Context {}

/**
 * This class is essentially a wrapper around fetch. It simplifies communication with the Mollie API over the network.
 */
export default class NetworkClient {
  /**
   * Triggers a request to the Mollie API.
   *
   * In contrast to the underlying `fetch` function, this function will:
   * - retry the request in some scenarios (see `retryingFetch`)
   * - throw an `ApiError` if the response from the Mollie API indicates an error
   * - appropriately process the response body before returning it (i.e. parsing it as JSON or throwing an ApiError if the response status indicates an error)
   */
  protected readonly request: (pathname: string, options?: RequestInit) => Promise<any>;
  constructor({
    apiKey,
    accessToken,
    versionStrings,
    apiEndpoint = 'https://api.mollie.com:443/v2/',
    caCertificates,
    libraryVersion,
    nodeVersion,
  }: Options & { caCertificates?: SecureContextOptions['ca']; libraryVersion: string; nodeVersion: string }) {
    // Compose the headers set in the sent requests.
    const headers: Record<string, string> = {};
    headers['User-Agent'] = composeUserAgent(nodeVersion, libraryVersion, versionStrings);
    if (apiKey != undefined) {
      headers['Authorization'] = `Bearer ${apiKey}`;
    } /* if (accessToken != undefined) */ else {
      headers['Authorization'] = `Bearer ${accessToken}`;
      headers['User-Agent'] += ' OAuth/2.0';
    }
    headers['Accept'] = 'application/hal+json';
    headers['Accept-Encoding'] = 'gzip';
    headers['Content-Type'] = 'application/json';

    // Create the https agent.
    const agent = new https.Agent({ ca: caCertificates });

    // Create retrying fetch function.
    const fetchWithRetries = retryingFetch(fetch);

    // normalize the API endpoint - if you provided a path, you probably meant to include it in the full URL. Without a trailing slash it would be ignored.
    if (!apiEndpoint.endsWith('/')) {
      apiEndpoint += '/';
    }

    // Create the request function.
    this.request = (pathname, options) => {
      const url = new URL(pathname, apiEndpoint);
      return fetchWithRetries(url, { agent, ...options, headers: { ...headers, ...options?.headers } })
        .catch(throwApiError)
        .then(processFetchResponse);
    };
  }

  async post<R>(pathname: string, data: Data & IdempotencyParameter, query?: SearchParameters): Promise<R | true> {
    // Take the idempotency key from the data, if any. It would be cleaner from a design perspective to have the
    // idempotency key in a separate argument instead of cramming it into the data like this. However, having a
    // separate argument would require every endpoint to split the input into those two arguments and thus cause a lot
    // of boiler-plate code.
    const { idempotencyKey, ...body } = data;
    const config: RequestInit = {
      method: 'POST',
      headers: idempotencyKey ? { [idempotencyHeaderName]: idempotencyKey } : undefined,
      body: JSON.stringify(body),
    };
    return this.request(buildUrl(pathname, query), config);
  }

  async get<R>(pathname: string, query?: SearchParameters): Promise<R> {
    return this.request(buildUrl(pathname, query));
  }

  async list<R>(pathname: string, binderName: string, query?: SearchParameters): Promise<R[]> {
    const data = await this.request(buildUrl(pathname, query));
    try {
      /* eslint-disable-next-line no-var */
      var { _embedded: embedded } = data;
    } catch (error) {
      throw new ApiError('Received unexpected response from the server');
    }
    return embedded[binderName] as R[];
  }

  async page<R>(pathname: string, binderName: string, query?: SearchParameters): Promise<R[] & Pick<Page<R>, 'links'>> {
    const data = await this.request(buildUrl(pathname, query));
    try {
      /* eslint-disable-next-line no-var */
      var { _embedded: embedded, _links: links } = data;
    } catch (error) {
      throw new ApiError('Received unexpected response from the server');
    }
    return Object.assign(embedded[binderName] as R[], { links });
  }

  iterate<R>(pathname: string, binderName: string, query: Maybe<SearchParameters>, valuesPerMinute = 5000): HelpfulIterator<R> {
    return new DemandingIterator(demand => {
      // Pick the page sizes (limits) based on the guessed demand.
      let popLimit: () => number;
      // If no demand was guessed, pages of 128 values each will be requested.
      if (demand == Number.POSITIVE_INFINITY) {
        popLimit = () => 128;
      } /* if (demand != Number.POSITIVE_INFINITY) */ else {
        let pageSizes: number[];
        // If a demand of 64 or less was guessed, request an initial page of 64 values. This ensures inaccurate guesses
        // do not result in wastefully short pages.
        if (demand <= 64) {
          pageSizes = [64];
        } /* if (demand > 64 && demand != Number.POSITIVE_INFINITY) */ else {
          // If a demand of over 64 was guessed, request pages of 250 values ‒ a limit imposed by the Mollie API ‒
          // until the remaining demand can be met with a single page. Finally, request that single page.
          pageSizes = new Array(Math.ceil(demand / 250));
          pageSizes[0] = demand - (pageSizes.length - 1) * 250;
          pageSizes.fill(250, 1);
        }
        // If more values are requested after the guessed demand (the guess was inaccurate), request pages of 128
        // values each.
        popLimit = () => pageSizes.pop() ?? 128;
      }
      // Set up the throttler. The iterator will wait before making a request to the Mollie API if more values have
      // been consumed than "allowed" by the valuesPerMinute argument. Note that this iterator will not interrupt the
      // yielding of values which have already been received from the Mollie API. If a page of 250 values is received
      // and valuesPerMinute is set to 100, all 250 values received values will be yielded before the (two-minute)
      // break.
      const throttler = new Throttler(valuesPerMinute);
      const { request } = this;
      return new HelpfulIterator<R>(
        (async function* iterate<R>() {
          let url = buildUrl(pathname, { ...query, limit: popLimit() });
          while (true) {
            // Request and parse the page from the Mollie API.
            const data = await request(url);
            try {
              /* eslint-disable-next-line no-var */
              var { _embedded: embedded, _links: links } = data;
            } catch (error) {
              throw new ApiError('Received unexpected response from the server');
            }
            // Yield the values on the page.
            const values = embedded[binderName] as R[];
            yield* values;
            // Inform the throttler of the yielded values.
            throttler.tally(values.length);
            // If the end of the sequence was reached ‒ in other words: this was the last page ‒ break out of the loop.
            if (links.next == null) {
              break;
            }
            // Build a URL from the "next" link in the response.
            const [pathname, query] = breakUrl(links.next.href);
            url = buildUrl(pathname, { ...query, limit: popLimit() });
            // Apply throttling.
            await throttler.throttle();
          }
        })(),
      );
    });
  }

  async patch<R>(pathname: string, data: Data): Promise<R> {
    const config: RequestInit = {
      method: 'PATCH',
      body: JSON.stringify(data),
    };
    return this.request(buildUrl(pathname), config);
  }

  async delete<R>(pathname: string, context?: Context & IdempotencyParameter): Promise<R | true> {
    // Take the idempotency key from the context, if any.
    const { idempotencyKey, ...body } = context ?? {};
    const config: RequestInit = {
      method: 'DELETE',
      headers: idempotencyKey ? { [idempotencyHeaderName]: idempotencyKey } : undefined,
      body: JSON.stringify(body),
    };
    return this.request(buildUrl(pathname), config);
  }
}
