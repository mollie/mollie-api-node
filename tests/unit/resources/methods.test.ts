import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import MethodsBinder from '../../../src/binders/methods/MethodsBinder';
import NetworkClient from '../../../src/communication/NetworkClient';

import response from '../__stubs__/methods.json';
import ApiError from '../../../src/errors/ApiError';

const mock = new MockAdapter(axios);

describe('methods', () => {
  let methods: MethodsBinder;
  beforeEach(() => {
    methods = new MethodsBinder(new NetworkClient({ apiKey: 'mock-api-key', nodeVersion: process.version, libraryVersion: 'mock' }));
  });

  describe('.get()', () => {
    const methodId = 'ideal';
    const error = { detail: 'The method id is invalid' };

    mock.onGet(`/methods/${methodId}`).reply(200, response._embedded.methods[0], {});
    mock.onGet('/methods/foo').reply(500, error, {});

    it('should return a method instance', () => 
      methods.get(methodId).then(result => {
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      methods.get(methodId, {}, (err, result) => {
        expect(err).toBeNull();
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should throw an error for non-existent IDs', () => 
      methods
        .get('foo')
        .then(result => expect(result).toBeUndefined())
        .catch(err => {
          expect(err).toBeInstanceOf(ApiError);
          expect(err.getMessage()).toEqual(error.detail);
        }));

    it('should return an error with a callback for non-existent IDs', done => {
      methods.get('foo', {}, (err: any, result) => {
        expect(err).toBeInstanceOf(ApiError);
        expect(err.getMessage()).toEqual(error.detail);
        expect(result).toBeUndefined();
        done();
      });
    });
  });

  describe('.list()', () => {
    mock.onGet('/methods').reply(200, response);

    it('should return a list of all methods', () =>
      methods.list().then(result => {
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      methods.list({}, (err, result) => {
        expect(err).toBeNull();
        expect(result).toMatchSnapshot();
        done();
      });
    });
  });
});
