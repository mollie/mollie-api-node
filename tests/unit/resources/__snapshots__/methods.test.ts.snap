// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`methods .get() should return a method instance 1`] = `
{
  "_links": {
    "self": {
      "href": "https://api.mollie.com/v2/methods/ideal",
      "type": "application/hal+json",
    },
  },
  "description": "iDEAL",
  "id": "ideal",
  "image": {
    "size1x": "https://www.mollie.com/images/payscreen/methods/ideal.png",
    "size2x": "https://www.mollie.com/images/payscreen/methods/ideal%402x.png",
  },
  "resource": "method",
}
`;

exports[`methods .get() should work with a callback 1`] = `
{
  "_links": {
    "self": {
      "href": "https://api.mollie.com/v2/methods/ideal",
      "type": "application/hal+json",
    },
  },
  "description": "iDEAL",
  "id": "ideal",
  "image": {
    "size1x": "https://www.mollie.com/images/payscreen/methods/ideal.png",
    "size2x": "https://www.mollie.com/images/payscreen/methods/ideal%402x.png",
  },
  "resource": "method",
}
`;

exports[`methods .list() should return a list of all methods 1`] = `
[
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/ideal",
        "type": "application/hal+json",
      },
    },
    "description": "iDEAL",
    "id": "ideal",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/ideal.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/ideal%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/creditcard",
        "type": "application/hal+json",
      },
    },
    "description": "Credit card",
    "id": "creditcard",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/creditcard.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/creditcard%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/banktransfer",
        "type": "application/hal+json",
      },
    },
    "description": "Bank transfer",
    "id": "banktransfer",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/banktransfer.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/banktransfer%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/inghomepay",
        "type": "application/hal+json",
      },
    },
    "description": "ING Home'Pay",
    "id": "inghomepay",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/inghomepay.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/inghomepay%402x.png",
    },
    "resource": "method",
  },
]
`;

exports[`methods .list() should work with a callback 1`] = `
[
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/ideal",
        "type": "application/hal+json",
      },
    },
    "description": "iDEAL",
    "id": "ideal",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/ideal.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/ideal%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/creditcard",
        "type": "application/hal+json",
      },
    },
    "description": "Credit card",
    "id": "creditcard",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/creditcard.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/creditcard%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/banktransfer",
        "type": "application/hal+json",
      },
    },
    "description": "Bank transfer",
    "id": "banktransfer",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/banktransfer.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/banktransfer%402x.png",
    },
    "resource": "method",
  },
  {
    "_links": {
      "self": {
        "href": "https://api.mollie.com/v2/methods/inghomepay",
        "type": "application/hal+json",
      },
    },
    "description": "ING Home'Pay",
    "id": "inghomepay",
    "image": {
      "size1x": "https://www.mollie.com/images/payscreen/methods/inghomepay.png",
      "size2x": "https://www.mollie.com/images/payscreen/methods/inghomepay%402x.png",
    },
    "resource": "method",
  },
]
`;
